### PR TITLE
Create Skip Connections and Branching

### PR DESCRIPTION
**Pull request by wmatray**
- Skip connections
- Branching (2 new nodes + skip connection)

Requires #2 

#

### CODE REVIEW

File: src/core.py
Comment (MortimerE) [inline src/core.py:1]: We/I can refactor this later but I kind of feel like we could change to doing class imports instead of a decision tree for the add_node function, mostly for style at this point but I think for functionality down the line
the elifs are going to get really long if we keep adding node types, which it sounds like we are 
I'm thinking we could have a function for each in an add_node class and then pass a list of operations to the evolver's mutation step which includes all of the different functions from add node plus other stuff like delete etc 
Then we randomly or weighted random select from that list and pass it to the mutation operation as a callback

Comment (wmatray) [reply src/core.py:1]: No problem with me here, I'm just building out and testing adding all the different types of mutations

File: src/core.py [hunk 1]
```diff
@@ -93,6 +93,44 @@ def add_node(graph: torch.fx.GraphModule, reference_node: torch.fx.Node, operati
...
+        # We then add a skip connection between the two new nodes
+        # Then we connect the skip connection to the users of the reference node
+        # y = f(x) -> y = h(g(x), k(x))
+        
+        # Get the shape of the reference node from metadata
+        reference_node_shape = reference_node.meta['tensor_meta'].shape
+        
+        # Create two new nodes with random shapes
+        branch1_shape = (reference_node_shape[-1], random.randint(1, 1000))
+        branch2_shape = (reference_node_shape[-1], random.randint(1, 1000))
```
[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/core.py:122]: Also, why is the last dim (`reference_node_shape[-1]`) becoming the first one? Didn't realize this at first, I thought it was just the batch dim

File: src/utils.py [hunk 2]
```diff
@@ -102,3 +102,151 @@ def add_specific_node(graph, reference_node, module):
...
+    if branch1_shape[-1] != branch2_shape[-1]:
+        target_size = max(branch1_shape[-1], branch2_shape[-1])
+        
+        # Adapt first branch if needed
+        if branch1_shape[-1] != target_size:
+            graph, final_branch1_node = adapt_node_shape(graph, branch1_node, branch1_shape[-1], target_size)
+        
+        # Adapt second branch if needed
+        if branch2_shape[-1] != target_size:
+            graph, final_branch2_node = adapt_node_shape(graph, branch2_node, branch2_shape[-1], target_size)
```
Comment (PatrickNercessian) [inline src/utils.py:226]:
```suggestion
        # Adapt first branch if needed
        if branch1_shape[-1] != target_size:
            graph, final_branch1_node = adapt_node_shape(graph, branch1_node, branch1_shape[-1], target_size)
        else:  # Adapt second branch if needed
            graph, final_branch2_node = adapt_node_shape(graph, branch2_node, branch2_shape[-1], target_size)
```

File: src/utils.py [hunk 3]
```diff
@@ -102,3 +102,151 @@ def add_specific_node(graph, reference_node, module):
...
+    # Infer the shapes of the branch nodes from the metadata
+    branch1_shape = branch1_node.meta['tensor_meta'].shape
+    branch2_shape = branch2_node.meta['tensor_meta'].shape
+    
+    # Initialize variables to track the final nodes to use in skip connection
+    final_branch1_node = branch1_node
+    final_branch2_node = branch2_node
+    
+    # Adapt branch nodes if needed to ensure they have compatible shapes
+    if branch1_shape[-1] != branch2_shape[-1]:
```
Comment (PatrickNercessian) [inline src/utils.py:217]: Would it be better to just have their shape be the same random number from the start? Otherwise we will always add one adaptive node in order to branch.

The branches can always have their shape modified later, they won't be locked into place at the same shape of course.

I am not sure if this is any better, just a thought

File: src/core.py [hunk 4]
```diff
@@ -156,39 +202,49 @@ def adapt_connections(
...
+        if first_shape[-1] != second_shape[-1]:
+            target_size = max(first_shape[-1], second_shape[-1])
+            
+            # Adapt first node if needed
+            if first_shape[-1] != target_size:
+                graph, first_node = adapt_node_shape(graph, first_node, first_shape[-1], target_size)
+            
+            # Adapt second node if needed
+            if second_shape[-1] != target_size:
+                graph, second_node = adapt_node_shape(graph, second_node, second_shape[-1], target_size)
```
Comment (PatrickNercessian) [inline src/core.py:230]: I think this is the same logic as in utils.py branch function, should we abstract it to a function? Not sure if the next two lines can also be included (which would let you include the if statement too).

**Review by PatrickNercessian — CHANGES_REQUESTED** [review]
I love it. Everything is minor changes or just questions for my curiosity

File: src/utils.py [hunk 5]
```diff
@@ -102,3 +102,151 @@ def add_specific_node(graph, reference_node, module):
...
+    Returns:
+        graph: The modified graph
+        new_node: The combined node after the branches
+    """
+    # Generate unique names for the branch nodes
+    branch1_name = get_unique_name(graph, "branch1")
+    branch2_name = get_unique_name(graph, "branch2")
+    
+    # Get the shape of the reference node from metadata
+    reference_node_shape = reference_node.meta['tensor_meta'].shape
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/utils.py:185]: Sorry I actually resolved this in my latest PR. This is an error in how I used the shapeprop as shapeprop should always start with the graph's input, called "placeholder" in the graph. In this case we're creating a shape based on this node.

File: src/utils.py [hunk 6]
```diff
@@ -102,3 +102,151 @@ def add_specific_node(graph, reference_node, module):
...
+        if branch2_shape[-1] != target_size:
+            graph, final_branch2_node = adapt_node_shape(graph, branch2_node, branch2_shape[-1], target_size)
+
+    # Run shape propagation to update metadata for the branch nodes
+    example_input = torch.randn(reference_node_shape)
+    try:
+        ShapeProp(graph).propagate(example_input)
+    except Exception as e:
+        print(f"Error during shape propagation: {e}")
+        print(f"Graph: {graph}")
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/utils.py:234]: Yeah this was for my debugging, I'll remove it

File: src/core.py [hunk 7]
```diff
@@ -93,6 +93,44 @@ def add_node(graph: torch.fx.GraphModule, reference_node: torch.fx.Node, operati
...
+        
+        # Create the branch modules
+        branch1_module = nn.Linear(branch1_shape[0], branch1_shape[1])
+        branch2_module = nn.Linear(branch2_shape[0], branch2_shape[1])
+        
+        # Use the utility function to add branch nodes
+        graph, new_node, skip_connection_output_shape = add_branch_nodes(graph, reference_node, branch1_module, branch2_module)
+        
+        # Determine the shape of the new node
+        new_node_shape = skip_connection_output_shape
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py:128]: Committed

[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py:128]: Oh ya commented on the wrong thread

File: src/core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py]: Yes, linear layers always apply to the last dimension. The linear layer is matrix multiplication (+ a bias) so for example:

(32 batch, 8 heads, 100 seq length, 32 features) X (32 features, 64 features) = (32 batch, 8 heads, 100 seq length, 64 features)

File: src/utils.py (hunk 2 above)
Comment (wmatray) [reply src/utils.py:226]: I fixed this in my next PR, to always just adapt the first branch to the second if need, no point in adding unnecessary adaptation nodes

File: src/core.py (hunk 4 above)
Comment (wmatray) [reply src/core.py:224]: I agree with you, fixed in core.py. We'll leave the code in just in case we change our minds about initializing them with the same shape, but as it stands the adaption checks will always fail and be ignored.

Comment (PatrickNercessian) [reply src/core.py:223]: I think you commented on the wrong comment here, but cool sounds good. First comment here still stands tho

File: src/utils.py (hunk 2 above)
Comment (PatrickNercessian) [reply src/utils.py:226]: I'm fine with it being in the next PR, but my suggested change here actually keeps functionality exactly the same (the current way you have it always has one or the other if-statement run, so I just suggested using an else)

**Comment by PatrickNercessian** [comment]
I think [this commit](https://github.com/PatrickNercessian/evolving-dnn/pull/4/commits/d1a1e67fd63bc4fee2142ee9b6c09e55c83c316b) added a bunch of Files Changed in. I think I have seen this before with Mort's merges, it's possibly a GitHub bug. The only way to remove it so we can make sure there's not unwanted changes in there would be to either close this PR and open another one (same branch) or just delete the last commit and force push, and then just do "Resolve conflicts" in the GitHub UI

File: src/core.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/core.py]: This one still uncommitted I think (might need to be done manually now since it's `Outdated`)

File: src/utils.py (hunk 2 above)
Comment (PatrickNercessian) [reply src/utils.py:226]: In other words, I think you're already only doing one adaptation

File: src/core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/core.py]: Gotcha.

For this one though:

> Also, why is the last dim (reference_node_shape[-1]) becoming the first one? Didn't realize this at first, I thought it was just the batch dim

I think I get it now: when you initialize the nn.Linear you don't need the other dimensions, and that's handled downstream when you do the shapeprop stuff. Is that right?

File: src/core.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py]: I made the change along with another edit to the same lines, the change is in there now

File: src/core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py]: Right the shapeprop tracks all the dimensions. For a linear layer all that matters is the last dimension.

File: src/core.py (hunk 4 above)
Comment (wmatray) [reply src/core.py:224]: Yes you're right, wrong comment thread!

> I think this is the same logic as in utils.py branch function, should we abstract it to a function? Not sure if the next two lines can also be included (which would let you include the if statement too).

```suggestion
            target_size = max(first_shape[-1], second_shape[-1])
            
            # Adapt first node if needed
            graph, first_node = adapt_node_shape(graph, first_node, first_shape[-1], target_size)
            
            # Adapt second node if needed
            graph, second_node = adapt_node_shape(graph, second_node, second_shape[-1], target_size)
```

You're right the check would happen twice, we can leave the check in the adapt_node_shape function.

File: src/utils.py (hunk 3 above)
Comment (wmatray) [reply src/utils.py:217]: Meant to comment here:

I agree with you, fixed in core.py. We'll leave the code in just in case we change our minds about initializing them with the same shape, but as it stands the adaption checks will always fail and be ignored.

File: src/core.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/core.py]: It's not :(

**Review by PatrickNercessian — APPROVED** [review]
Not important changes left, so approving! Down to merge now or after you change accordingly to these 2 comments.

File: src/core.py (hunk 4 above)
Comment (wmatray) [reply src/core.py]: There's no need to have the redundant check in core.py. Removed in latest commit.

File: src/core.py [hunk 8]
```diff
@@ -104,7 +136,11 @@ def add_node(graph: torch.fx.GraphModule, reference_node: torch.fx.Node, operati
 
     # Run shape propagation again to update all shape metadata
     example_input = torch.randn(next(iter(graph.graph.nodes)).meta['tensor_meta'].shape)
-    ShapeProp(graph).propagate(example_input)
+    try:
+        ShapeProp(graph).propagate(example_input)
+    except Exception as e:
+        print(f"Error during shape propagation: {e}")
+        print(f"Graph: {graph}")
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py]: Removed in latest commit

**Review by MortimerE — APPROVED** [review]
Just followed up on everything, looks hot
